### PR TITLE
Additions

### PR DESCRIPTION
**Pull request by Kathleendeboer**

### CODE REVIEW

File: fairseq/tasks/fairseq_task.py [hunk 1]
```diff
@@ -537,10 +537,17 @@ def train_step(
...
         with torch.no_grad():
-            loss, sample_size, logging_output = criterion(model, sample)
-        return loss, sample_size, logging_output
+            logging_output = criterion(model, sample)  #calling the same criterion loss function #loss deleted MODIFIED
+            #print("Prediction", sample_size) #MODIFIED
+            #print("Probabilities", logging_output)
+            preds = list(logging_output)[0]  # First value
+            probabilities = list(logging_output)[1]  # Second value
+
+        return preds, probabilities
```
Comment (Veldhoen) [inline fairseq/tasks/fairseq_task.py:549]: You are modifying the blueprint of the function here, which affects the compatibility. Is that really necessary, or could you also create a new function with your desired return type?

File: fairseq/tasks/fairseq_task.py [hunk 2]
```diff
@@ -537,10 +537,17 @@ def train_step(
         return loss, sample_size, logging_output
 
     def valid_step(self, sample, model, criterion):
+   
         model.eval()
         with torch.no_grad():
-            loss, sample_size, logging_output = criterion(model, sample)
-        return loss, sample_size, logging_output
+            logging_output = criterion(model, sample)  #calling the same criterion loss function #loss deleted MODIFIED
+            #print("Prediction", sample_size) #MODIFIED
```
Comment (Veldhoen) [inline fairseq/tasks/fairseq_task.py:544]: Remove spurious print comments (old print statements)

File: fairseq/tasks/emotion_prediction.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (Veldhoen) [inline fairseq/tasks/emotion_prediction.py:1]: Is this Facebook code or your own?

File: fairseq/tasks/emotion_prediction.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                                               data_split=split, 
+                                               sample_rate=self.args.sample_rate,
+                                               max_sample_size=self.args.max_sample_size,
+                                               min_sample_size=self.args.min_sample_size)
+
+    def build_model(self, args):
+        
+        from fairseq import models
+      
+        model = models.build_model(args, self) #/home/gsir059/Documents/PhD/MulFie/fairseq/fairseq/models/__init__.py'
```
Comment (Veldhoen) [inline fairseq/tasks/emotion_prediction.py:111]: weird comment

File: fairseq/tasks/emotion_prediction.py
Comment (Veldhoen) [inline fairseq/tasks/emotion_prediction.py:1]: It seems like you based your version on an existing example file. It's more clear if you first commit theexample file, and then commit your changes to it. That makes it easier to review your additions/modifications. 
Could you point me to the file you used as a starting point? Or even better, try to insert it post hoc? See https://printf2linux.wordpress.com/2012/04/09/insert-a-commit-in-the-past-git/ for an instruction of inserting a commit in your history

File: sseft_shap.py [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+import matplotlib.colors as mcolors
+
+## This file contains the explainability framework merged with the model 'sse-ft', it is a work in progress, however the masking logic is tested and is correct
+
+## Main for now parses the args from the command prompt and passes them to load_models()
+
+
+# Load_models() gets the model from the fairseq library and sets the checkpoint, it initialises a data iterator and loops over the samples in the dataset. 
+# For now I need to test the masking functions together with the model, later i will adjust the code so that it can handle data from Sound and Vision
+global text_tensor
```
Comment (Veldhoen) [inline sseft_shap.py:27]: You have a lot of global variables. In general, it is adviced to use scoped variables whenever possible, to avoid issues of overshadowing and make it easier to follow the flow of your code.

File: sseft_shap.py [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+    # Define the token IDs for Friends characters
+
+    token_to_word.update({7423: "Rachel"})
+    token_to_word.update({4129: "Phoebe"})
+    token_to_word.update({3540: "Phoebe"})
+    token_to_word.update({1610: "Phoebe"})
+    token_to_word.update({4012: "Ross"})
+    token_to_word.update({12811: "Monica"})
+    token_to_word.update({16027: "Chandler"})
+    token_to_word.update({12972: "Joey"})
```
Comment (Veldhoen) [inline sseft_shap.py:139]: This was added for testing, I presume? Replace this bit with something meaningful, or at least add a TODO statement!

File: sseft_shap.py
Comment (Veldhoen) [inline sseft_shap.py:1]: I wonder whether it makes sense to add this module to this repo, or rather in the worker itself. It doesn't seem to really be a part of fairseq, at first glance at least.

File: sseft_shap.py [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+        logger.info("Over all labels: TEXT_score: %s, AUDIO_score: %s, VIDEO_score: %s", text_score, audio_score, video_score)
+
+        
+    avg_text_score = total_text_score / num_samples # Normalize for all samples
+    avg_audio_score = total_audio_score / num_samples
+    avg_video_score = total_video_score / num_samples
+    # Print the average multimodality scores for the entire dataset for one label
+    logger.info("Average Text Score: %s", avg_text_score)
+    logger.info("Average Audio Score: %s", avg_audio_score)
+    logger.info("Average Video Score: %s", avg_video_score)
```
Comment (Veldhoen) [inline sseft_shap.py:242]: You probably want to add some return statement, or include code to output results to file, rather than just log this information.

File: sseft_shap.py [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+            #num_patches = 256
+            num_patches = 16
+            patch_size = 64
+            num_frames = 300
+            num_patches_per_row = 4
+            #Patches as tokens
+            for k in range(masked_video_tokens_ids.size(1)-1):
+                if masked_video_tokens_ids[i,k].item() == 0 :
+                    # start_row = (k // (256 // patch_size)) * patch_size
+                    # start_col = (k % (256 // patch_size)) * patch_size
```
Comment (Veldhoen) [inline sseft_shap.py:281]: Spurious comments

File: sseft_shap.py [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+                      if torch.any(masked_video_tensor[:,:,frame_idx,:,:]== -1):
+                        continue  # Skip padded frames
+                      masked_video_tensor[:,:, frame_idx, start_row:start_row+patch_size, start_col:start_col+patch_size] = 0 ## Here the real video masking happens
+            frames_per_token = video_depth // masked_video_tokens_ids.size(1)
+            #Timeframes as tokens
+            # for k in range((masked_video_tokens_ids.size(1)-1)):
+           #     start_frame = k * frames_per_token
+            #    end_frame = min((k + 1) * frames_per_token, video_depth)
+            #     if masked_video_tokens_ids[i,k].item() == 0 :
+            #         masked_video_tensor[:,:,start_frame:end_frame,:,:] = 0
```
Comment (Veldhoen) [inline sseft_shap.py:294]: Spurious comments

File: sseft_shap.py [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+## The custom masker recieves the input tokens and a boolean tensor of the same shape (what need to be masked)
+def custom_masker(mask, x):
+    masked_X = x.clone()
+    mask = torch.tensor(mask).unsqueeze(0)
+    masked_X[~mask] = 0  # ~mask !!! to zero
+    # never mask out CLS and SEP tokens (makes no sense for the model to work without them) -> this is handled in the load_models range(1,...) instead of range*(0,....)
+    return masked_X
+
+## This function recieves an array with rows that represent combinations of masked tokens for one sample
+## From the masked tokens, based on the token generaton logic, the input is reconstructed to the form needed by the model
```
Comment (Veldhoen) [inline sseft_shap.py:253]: Please add a bit of explanation here or within the function. What is the difference between masked tokens and masked tensors, and what is the input 'x'?

File: sseft_shap.py [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+            masked_audio_tensor = audio_tensor.clone()
+            for k in range(masked_audio_tokens_ids.size(1)):
+                if masked_audio_tokens_ids[i,k].item() == 0:
+                    masked_audio_tensor[k:k+samples_per_token] = 0 ## Here the real audio masking happens
+            masked_video_tensor = video_tensor.clone()
+            #num_patches = 256
+            num_patches = 16
+            patch_size = 64
+            num_frames = 300
+            num_patches_per_row = 4
```
Comment (Veldhoen) [inline sseft_shap.py:276]: Should these be hard-coded or an argument?

File: sseft_shap.py [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+    return result
+#if torch.any(masked_video_tensor[0, :, frame_idx] == -1): TEST
+def compute_mm_score_local(shap_values):
+    """ Compute Multimodality Score. . """
+    text_contrib = np.abs(shap_values.values[0, 0, :text_length]).sum()/text_length 
+    audio_contrib = np.abs(shap_values.values[0, 0, text_length: text_length + audio_length]).sum()/audio_length
+    video_contrib = np.abs(shap_values.values[0, 0, text_length + audio_length:]).sum()/16
+    # text_contrib = np.abs(shap_values.values[:, :, :text_length]).sum(axis=(0, 1)) / text_length
+    # audio_contrib = np.abs(shap_values.values[:, :, text_length: text_length + audio_length]).sum(axis=(0, 1))/audio_length
+    # video_contrib = np.abs(shap_values.values[:, :, text_length + audio_length:]).sum(axis=(0,1))/16
```
Comment (Veldhoen) [inline sseft_shap.py:332]: Spurious comments

File: fairseq/modules/transformer_layer.py
Comment (Veldhoen) [inline fairseq/modules/transformer_layer.py:1]: I think this is the only place where you really adjust the code, rather than add your own tasks. The modifications seem quite big. Should you not create a new module for you layer, or is it right to replace the existing transformer_layer? I'm asking because of compatibility and staying in sync with the original code.

File: fairseq/models/mulT_emo.py
Comment (Veldhoen) [inline fairseq/models/mulT_emo.py:1]: For this file as well, I'd like to be able to compare with a base/example that you based your work on

File: fairseq/data/raw_audio_text_video_dataset.py [hunk 13]
```diff
@@ -0,0 +1,398 @@
+#change the line 331
```
Comment (Veldhoen) [inline fairseq/data/raw_audio_text_video_dataset.py:1]: ???

File: fairseq/data/raw_audio_text_video_dataset.py [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+
+                    if self.data_args.eval_matric:
+                        if emotion==0.0:
+                            continue  
+
+                    self.fnames_audio.append(items_a[0])
+                    self.fnames_text.append(items_t[0])
+                    self.fnames_video.append(items_v[0])
+                    self.sizes_audio.append(1000000)     #This is used in the data loader np.lexsort but can remove it
+                    self.sizes_video.append(1000000)
```
Comment (Veldhoen) [inline fairseq/data/raw_audio_text_video_dataset.py:127]: Hardcoded number always make me suspicious

File: fairseq/data/raw_audio_text_video_dataset.py [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+
+            
+
+                if self.data_args.regression_target_mos:
+
+                    if self.data_args.eval_matric:
+                        if emotion==0.0:
+                            continue  
+
+                    self.fnames_audio.append(items_a[0])
```
Comment (Veldhoen) [inline fairseq/data/raw_audio_text_video_dataset.py:123]: Why just append one item here?

File: fairseq/data/raw_audio_text_video_dataset.py [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+                '1':1,
+                '2':2,
+                '3':3
+            }
+
+            # self.emotion_dictionary = {   #modei senti 2 class
+            
+            #     '0': 0,
+            #     '1':1
+            # }
```
Comment (Veldhoen) [inline fairseq/data/raw_audio_text_video_dataset.py:193]: Alternative encoding is commented out? Should this be configurable?

File: fairseq/data/raw_audio_text_video_dataset.py [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+    def collater(self, samples):
+        
+        if len(samples) == 0:
+            return {}
+        #collater for audio data    
+        ###################################################################    
+        sources = [s['audio'] for s in samples]
+        #print("DEBUG dataset sources length")
+        #print(len(sources))
+        #print(sources)
```
Comment (Veldhoen) [inline fairseq/data/raw_audio_text_video_dataset.py:306]: Spurious comments

File: fairseq/criterions/emotion_prediction_cri.py
Comment (Veldhoen) [inline fairseq/criterions/emotion_prediction_cri.py:1]: Here as well, please provide the basis that you extended explicitly.

Also, the functions are quite complex: refactor/break them down into meaningful, testable bits. Provide a description (ideally docstring) explaining what each part should be doing.

File: fairseq/criterions/emotion_prediction_cri.py [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+                'sample_size': sample_size,
+                'ncorrect':ncorrect
+            }
+
+            # pred_real_i='pred_mos_real'
+            # truth_real_i='truth_mos_real'
+
+            # logging_output.update({truth_real_i : targets.view(-1).cpu().detach().numpy()})
+            # logging_output.update({pred_real_i : logits.view(-1).cpu().detach().numpy()})
+
```
Comment (Veldhoen) [inline fairseq/criterions/emotion_prediction_cri.py:93]: Spurious comments

File: fairseq/criterions/emotion_prediction_cri.py [hunk 19]
```diff
@@ -0,0 +1,388 @@
...
+
+            if self.args.eval_matric:  # For the binary
+
+                test_preds_np = logits.view(-1).cpu().detach().numpy()
+                test_truth_np= targets.view(-1).cpu().detach().numpy()
+
+                exclude_zero=True
+
+                #This gives a problem when running with with batch size of one and that batch consist of a '0' as the truth
```
Comment (Veldhoen) [inline fairseq/criterions/emotion_prediction_cri.py:102]: Not clear what you are trying to accomplish with exclude_zero

**Review by Veldhoen — CHANGES_REQUESTED** [review]
I've looked at the separate files now, and added quite some remarks. In general, I think your code needs some refactoring and cleaning up. 

I am not sure whether you really need to create this fairseq release, or whether you could also add your own functionality in the worker repo. Most items you added are additions rather than modifications, except the fairseq/modules/transformer_layer, in which your modifications are so impactful that it seems like you are actually creating new/different objects. 

I did not try to run your code yet. Please provide a description of what/how to run when you create a PR :)
